File: typescript/src/serializer.ts
/**
 * The `json` array->bytes codec for zarrita: chunks as canonical JSON arrays.
 *
 * Encoding is the Zarr v3 fill_value scalar serialization, applied elementwise
 * and nested by shape in C order: numbers as JSON numbers, NaN/Infinity as the
 * strings "NaN"/"Infinity"/"-Infinity", booleans as true/false. Output is
 * canonical JSON (see codec.canonicalStringify), which is what lets
 * ZarrJsonStore inline these chunks as real JSON arrays in the document.
 *
 * int64/uint64 caveat: JS represents these as BigInt. Values within
 * Number.MAX_SAFE_INTEGER are emitted as JSON numbers; larger values throw —
 * the known int64-in-JS limitation, inherited from the fill_value convention
 * Integers beyond 2^53 round-trip losslessly as bigint via strictParse.
 *
 * Register with zarrita via `registerJsonCodec()`, then create arrays with
 * `codecs: [{ name: "json", configuration: {} }]`.
 */

import { BoolArray, registry } from "zarrita";
import type { Chunk, CodecMetadata, DataType, TypedArray } from "zarrita";
import { canonicalStringify, strictParse } from "./codec.js";

type ChunkMeta = {
	dataType: DataType;
	shape: number[];
	codecs: CodecMetadata[];
	fillValue: unknown;
};

export function cStrides(shape: number[]): number[] {
	const stride = new Array<number>(shape.length);
	let step = 1;
	for (let i = shape.length - 1; i >= 0; i--) {
		stride[i] = step;
		step *= shape[i];
	}
	return stride;
}

function isFloatType(dataType: string): boolean {
	return (
		dataType === "float16" || dataType === "float32" || dataType === "float64"
	);
}

function isBigintType(dataType: string): boolean {
	return dataType === "int64" || dataType === "uint64";
}

/** Serialize one element per the Zarr v3 fill_value scalar convention. */
export function toJsonScalar(value: unknown, dataType: string): unknown {
	if (typeof value === "bigint") {
		// Safe values serialize as plain numbers (matching the other
		// implementations' output for small ints); larger values stay bigint
		// and canonicalStringify emits their exact digits.
		if (
			value > BigInt(Number.MAX_SAFE_INTEGER) ||
			value < -BigInt(Number.MAX_SAFE_INTEGER)
		) {
			return value;
		}
		return Number(value);
	}
	if (typeof value === "number") {
		if (Number.isNaN(value)) return "NaN";
		if (value === Number.POSITIVE_INFINITY) return "Infinity";
		if (value === Number.NEGATIVE_INFINITY) return "-Infinity";
		return value;
	}
	if (typeof value === "boolean" || typeof value === "string") {
		return value;
	}
	throw new Error(
		`json codec: cannot serialize ${dataType} scalar ${String(value)}`,
	);
}

const INT_RANGES: Record<string, [number, number]> = {
	int8: [-128, 127],
	int16: [-32768, 32767],
	int32: [-2147483648, 2147483647],
	uint8: [0, 255],
	uint16: [0, 65535],
	uint32: [0, 4294967295],
};

/**
 * Parse one element per the Zarr v3 fill_value scalar convention.
 *
 * Strict, like zarr-python's from_json_scalar: coercing silently would be
 * corruption. int64/uint64 accept bigint (strictParse yields it for integer
 * literals beyond 2^53) or safe-integer numbers, range-checked against the
 * dtype bounds; a plain number that is integral but unsafe is rejected (it
 * can only mean a lossy parse upstream). Other int dtypes reject
 * non-integers and out-of-range values; floats accept numbers plus the
 * "NaN"/"Infinity"/"-Infinity" strings; bool requires a JSON boolean.
 */
export function fromJsonScalar(value: unknown, dataType: string): unknown {
	if (isFloatType(dataType)) {
		if (typeof value === "number") {
			return value;
		}
		if (typeof value === "bigint") {
			// An integer token for a float type: the nearest float64 (SPEC §9.2).
			return Number(value);
		}
		if (value === "NaN") return Number.NaN;
		if (value === "Infinity") return Number.POSITIVE_INFINITY;
		if (value === "-Infinity") return Number.NEGATIVE_INFINITY;
		throw new Error(`json codec: invalid ${dataType} scalar ${JSON.stringify(value)}`);
	}
	if (isBigintType(dataType)) {
		// Integer types require an INTEGER token. Callers parse chunk JSON
		// with integersAsBigInt so every integer literal arrives as bigint;
		// a JS number here therefore means a float token like 1.0, which
		// SPEC §9.2 forbids for integer types even when integral.
		if (typeof value !== "bigint") {
			throw new Error(
				`json codec: invalid ${dataType} scalar ${String(value)}: ` +
					"must be an integer token",
			);
		}
		const big = value;
		const [lo, hi] =
			dataType === "int64"
				? [-(2n ** 63n), 2n ** 63n - 1n]
				: [0n, 2n ** 64n - 1n];
		if (big < lo || big > hi) {
			throw new Error(
				`json codec: ${dataType} scalar ${big} is out of range [${lo}, ${hi}]`,
			);
		}
		return big;
	}
	if (dataType === "bool") {
		if (typeof value !== "boolean") {
			throw new Error(`json codec: invalid ${dataType} scalar ${JSON.stringify(value)}`);
		}
		return value;
	}
	const range = INT_RANGES[dataType];
	if (range !== undefined) {
		// As for int64/uint64: an integer token (bigint) is required.
		if (
			typeof value !== "bigint" ||
			value < BigInt(range[0]) ||
			value > BigInt(range[1])
		) {
			throw new Error(`json codec: invalid ${dataType} scalar ${String(value)}`);
		}
		return Number(value);
	}
	return value;
}

/** Nest a flat C-order scalar list by shape. */
export function nest(flat: unknown[], shape: number[]): unknown {
	if (shape.length === 0) {
		return flat[0];
	}
	if (shape.length === 1) {
		return flat;
	}
	const step = flat.length / shape[0];
	const out: unknown[] = [];
	for (let i = 0; i < shape[0]; i++) {
		out.push(nest(flat.slice(i * step, (i + 1) * step), shape.slice(1)));
	}
	return out;
}

/**
 * Flatten nested chunk JSON to a C-order scalar list, driven by shape.
 *
 * Shape-driven: a scalar's JSON form may itself be an array (a complex scalar
 * is a [re, im] pair), so recursion must stop at the last dimension, not at
 * non-array values.
 */
function flatten(nested: unknown, shape: number[]): unknown[] {
	if (shape.length === 0) {
		return [nested];
	}
	if (!Array.isArray(nested) || nested.length !== shape[0]) {
		throw new Error(
			`json codec: chunk JSON does not match chunk shape [${shape.join(",")}]`,
		);
	}
	if (shape.length === 1) {
		return [...nested];
	}
	const out: unknown[] = [];
	for (const sub of nested) {
		out.push(...flatten(sub, shape.slice(1)));
	}
	return out;
}

type NumberArrayCtor = {
	from(values: Iterable<number>): TypedArray<DataType>;
};

const NUMBER_CTORS: Record<string, NumberArrayCtor> = {
	int8: Int8Array as unknown as NumberArrayCtor,
	int16: Int16Array as unknown as NumberArrayCtor,
	int32: Int32Array as unknown as NumberArrayCtor,
	uint8: Uint8Array as unknown as NumberArrayCtor,
	uint16: Uint16Array as unknown as NumberArrayCtor,
	uint32: Uint32Array as unknown as NumberArrayCtor,
	float32: Float32Array as unknown as NumberArrayCtor,
	float64: Float64Array as unknown as NumberArrayCtor,
};

export function makeTypedArray(dataType: string, scalars: unknown[]): TypedArray<DataType> {
	if (isBigintType(dataType)) {
		const Ctor = dataType === "int64" ? BigInt64Array : BigUint64Array;
		return Ctor.from(scalars as bigint[]) as unknown as TypedArray<DataType>;
	}
	if (dataType === "bool") {
		return new BoolArray(scalars as boolean[]) as unknown as TypedArray<DataType>;
	}
	const Ctor = NUMBER_CTORS[dataType];
	if (Ctor === undefined) {
		throw new Error(`json codec: unsupported data type ${JSON.stringify(dataType)}`);
	}
	return Ctor.from(scalars as number[]);
}

const UTF8_ENCODER = new TextEncoder();
const UTF8_DECODER = new TextDecoder("utf-8", { fatal: true });

/** Indexed element access covering both TypedArrays and BoolArray-likes. */
function elementAt(data: unknown, index: number): unknown {
	if (typeof (data as { get?: unknown }).get === "function") {
		return (data as { get(i: number): unknown }).get(index);
	}
	return (data as ArrayLike<unknown>)[index];
}

/**
 * Read a chunk's elements in C order of `shape`, honoring `stride`.
 *
 * zarrita's Chunk carries an explicit stride: an array->array codec earlier
 * in the chain (e.g. transpose) hands the array->bytes codec a chunk whose
 * memory is not C-contiguous. Walking raw memory order would emit JSON that
 * nests wrongly by shape (verified divergence against zarr-python/zarrs), so
 * elements are gathered by logical index.
 */
export function chunkElements(chunk: {
	data: unknown;
	shape: number[];
	stride: number[];
}): unknown[] {
	const { data, shape, stride } = chunk;
	const rank = shape.length;
	const size = shape.reduce((a, b) => a * b, 1);
	const out = new Array<unknown>(size);
	const index = new Array<number>(rank).fill(0);
	for (let n = 0; n < size; n++) {
		let offset = 0;
		for (let dim = 0; dim < rank; dim++) {
			offset += index[dim] * stride[dim];
		}
		out[n] = elementAt(data, offset);
		for (let dim = rank - 1; dim >= 0; dim--) {
			index[dim] += 1;
			if (index[dim] < shape[dim]) {
				break;
			}
			index[dim] = 0;
		}
	}
	return out;
}

/** Array->bytes codec encoding chunks as canonical UTF-8 JSON arrays. */
export class JsonSerializer {
	readonly kind = "array_to_bytes" as const;
	#dataType: DataType;
	#shape: number[];

	constructor(meta: ChunkMeta) {
		this.#dataType = meta.dataType;
		this.#shape = meta.shape;
	}

	static fromConfig(config: unknown, meta: ChunkMeta): JsonSerializer {
		if (
			config !== undefined &&
			config !== null &&
			(typeof config !== "object" || Object.keys(config).length > 0)
		) {
			throw new Error(
				`json codec takes no configuration; got ${JSON.stringify(config)}`,
			);
		}
		return new JsonSerializer(meta);
	}

	encode(chunk: Chunk<DataType>): Uint8Array {
		// Walk elements in C order of chunk.shape via chunk.stride: a
		// transpose codec earlier in the chain hands over non-C-contiguous
		// memory, and raw-order iteration would nest the JSON wrongly.
		const flat = chunkElements(chunk).map((v) => toJsonScalar(v, this.#dataType));
		const text = canonicalStringify(nest(flat, chunk.shape));
		return UTF8_ENCODER.encode(text);
	}

	decode(bytes: Uint8Array): Chunk<DataType> {
		// integersAsBigInt preserves each element's number sort: integer
		// tokens arrive as bigint (any size), float tokens as number.
		const nested: unknown = strictParse(UTF8_DECODER.decode(bytes), {
			integersAsBigInt: true,
		});
		const flat = flatten(nested, this.#shape);
		const scalars = flat.map((v) => fromJsonScalar(v, this.#dataType));
		return {
			data: makeTypedArray(this.#dataType, scalars),
			shape: this.#shape,
			stride: cStrides(this.#shape),
		};
	}
}

type Registry = typeof registry;
type RegistryValue = Parameters<Registry["set"]>[1];

/**
 * Register the `json` codec with a zarrita codec registry (the global
 * `zarrita.registry` by default), under the codec name "json".
 */
export function registerJsonCodec(reg: Registry = registry): void {
	reg.set("json", (async () => JsonSerializer) as unknown as RegistryValue);
}

File: rust/src/serializer.rs
//! The `json` array->bytes codec: chunks as canonical JSON arrays.
//!
//! Encoding is the Zarr v3 fill_value scalar serialization, applied
//! elementwise and nested by shape in C order. Every Zarr v3 data type must
//! define a JSON scalar serialization to be representable in the `fill_value`
//! metadata field, so this codec covers every fixed-size dtype by
//! construction: NaN/Infinity become the strings "NaN"/"Infinity", complex
//! values become [re, im] pairs, and so on. zarrs exposes the serialization
//! as `DataTypeTraits::metadata_fill_value` / `fill_value_v3`.
//!
//! Output is canonical JSON (see [`crate::document::canonical_to_string`]),
//! which is what lets [`crate::ZarrInlineStore`] inline these chunks as real
//! JSON arrays in the document.
//!
//! A rank-0 chunk serializes to a bare JSON scalar; the store only inlines
//! arrays, so rank-0 chunks are stored base64-encoded.
//!
//! The codec registers itself with zarrs's compile-time plugin registry
//! (inventory) under the name `json`, so arrays whose metadata names the
//! codec can be opened without further setup.

use std::borrow::Cow;
use std::num::NonZeroU64;
use std::sync::Arc;

use serde_json::Value;
use zarrs::array::codec::api::{
    ArrayCodecTraits, ArrayToBytesCodecTraits, Codec, CodecError, CodecMetadataOptions,
    CodecOptions, CodecPluginV3, CodecTraits, CodecTraitsV3, PartialDecoderCapability,
    PartialEncoderCapability, RecommendedConcurrency,
};
use zarrs::array::{
    ArrayBytes, ArrayBytesRaw, BytesRepresentation, DataType, FillValue, FillValueMetadata,
};
use zarrs::metadata::v3::MetadataV3;
use zarrs::metadata::Configuration;
use zarrs::plugin::{PluginCreateError, ZarrVersion};

use crate::document::canonical_to_string;

/// The `json` array->bytes codec: encodes chunks as canonical UTF-8 JSON
/// arrays using the Zarr v3 fill_value scalar serialization elementwise.
#[derive(Debug, Clone, Copy, Default)]
pub struct JsonCodec;

impl JsonCodec {
    /// Create a new `json` codec.
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

zarrs::plugin::impl_extension_aliases!(JsonCodec, v3: "json");

// Compile-time registration with zarrs's V3 codec plugin registry.
inventory::submit! {
    CodecPluginV3::new::<JsonCodec>()
}

impl CodecTraitsV3 for JsonCodec {
    fn create(metadata: &MetadataV3) -> Result<Codec, PluginCreateError> {
        // The json codec takes no configuration (SPEC.md section 9); reject
        // unrecognized configuration members rather than ignoring them.
        if let Some(configuration) = metadata.configuration() {
            if !configuration.is_empty() {
                return Err(PluginCreateError::from(
                    "json codec takes no configuration",
                ));
            }
        }
        Ok(Codec::ArrayToBytes(Arc::new(JsonCodec)))
    }
}

/// Nest a flat C-order list of scalars by `shape`.
fn nest(flat: &[Value], shape: &[u64]) -> Value {
    match shape {
        [] => flat[0].clone(),
        [_] => Value::Array(flat.to_vec()),
        [dim, rest @ ..] => {
            let step = flat.len() / usize::try_from(*dim).unwrap();
            Value::Array(
                (0..usize::try_from(*dim).unwrap())
                    .map(|i| nest(&flat[i * step..(i + 1) * step], rest))
                    .collect(),
            )
        }
    }
}

/// Shape-driven flattening: a scalar's JSON form may itself be an array
/// (complex -> [re, im]), so recursion must stop at the last dimension, not
/// at non-array values.
fn flatten(nested: &Value, shape: &[u64], out: &mut Vec<Value>) -> Result<(), CodecError> {
    let mismatch = || CodecError::Other(format!("json codec: chunk JSON does not match chunk shape {shape:?}"));
    match shape {
        [] => {
            out.push(nested.clone());
            Ok(())
        }
        [dim, rest @ ..] => {
            let Value::Array(items) = nested else {
                return Err(mismatch());
            };
            if items.len() as u64 != *dim {
                return Err(mismatch());
            }
            if rest.is_empty() {
                out.extend(items.iter().cloned());
                Ok(())
            } else {
                for sub in items {
                    flatten(sub, rest, out)?;
                }
                Ok(())
            }
        }
    }
}

/// For float data types, whether the native-endian element bytes hold a
/// finite value. Non-float types (or unknown widths) are trivially finite.
fn float_element_is_finite(type_name: &str, element: &[u8]) -> bool {
    let f64_finite = |b: &[u8]| f64::from_ne_bytes(b.try_into().unwrap()).is_finite();
    let f32_finite = |b: &[u8]| f32::from_ne_bytes(b.try_into().unwrap()).is_finite();
    match (type_name, element.len()) {
        ("float64", 8) => f64_finite(element),
        ("float32", 4) => f32_finite(element),
        // IEEE binary16: exponent bits 0x7C00 all set => inf/NaN.
        ("float16", 2) => u16::from_ne_bytes(element.try_into().unwrap()) & 0x7C00 != 0x7C00,
        // bfloat16: exponent bits 0x7F80 all set => inf/NaN.
        ("bfloat16", 2) => u16::from_ne_bytes(element.try_into().unwrap()) & 0x7F80 != 0x7F80,
        // Complex types are two float lanes; both must be finite.
        ("complex128", 16) => f64_finite(&element[..8]) && f64_finite(&element[8..]),
        ("complex64", 8) => f32_finite(&element[..4]) && f32_finite(&element[4..]),
        _ => true,
    }
}

fn fixed_size(data_type: &DataType) -> Result<usize, CodecError> {
    data_type.fixed_size().ok_or_else(|| {
        CodecError::UnsupportedDataType(data_type.clone(), "json".to_string())
    })
}

impl CodecTraits for JsonCodec {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn configuration(
        &self,
        _version: ZarrVersion,
        _options: &CodecMetadataOptions,
    ) -> Option<Configuration> {
        Some(Configuration::default())
    }

    fn partial_decoder_capability(&self) -> PartialDecoderCapability {
        PartialDecoderCapability {
            partial_read: false,
            partial_decode: false,
        }
    }

    fn partial_encoder_capability(&self) -> PartialEncoderCapability {
        PartialEncoderCapability {
            partial_encode: false,
        }
    }
}

impl ArrayCodecTraits for JsonCodec {
    fn recommended_concurrency(
        &self,
        _shape: &[NonZeroU64],
        _data_type: &DataType,
    ) -> Result<RecommendedConcurrency, CodecError> {
        Ok(RecommendedConcurrency::new_maximum(1))
    }
}

impl ArrayToBytesCodecTraits for JsonCodec {
    fn into_dyn(self: Arc<Self>) -> Arc<dyn ArrayToBytesCodecTraits> {
        self as Arc<dyn ArrayToBytesCodecTraits>
    }

    fn encoded_representation(
        &self,
        _shape: &[NonZeroU64],
        _data_type: &DataType,
        _fill_value: &FillValue,
    ) -> Result<BytesRepresentation, CodecError> {
        Ok(BytesRepresentation::UnboundedSize)
    }

    fn encode<'a>(
        &self,
        bytes: ArrayBytes<'a>,
        shape: &[NonZeroU64],
        data_type: &DataType,
        _fill_value: &FillValue,
        _options: &CodecOptions,
    ) -> Result<ArrayBytesRaw<'a>, CodecError> {
        let element_size = fixed_size(data_type)?;
        let num_elements = shape.iter().map(|d| d.get()).product::<u64>();
        bytes.validate(num_elements, data_type)?;
        let raw = bytes.into_fixed()?;

        // Elementwise fill_value scalar serialization.
        let mut flat = Vec::with_capacity(usize::try_from(num_elements).unwrap());
        for element in raw.chunks_exact(element_size) {
            let fill_value = FillValue::new(element.to_vec());
            let metadata = data_type
                .metadata_fill_value(&fill_value)
                .map_err(|e| CodecError::Other(format!("json codec: {e}")))?;
            let value = serde_json::to_value(&metadata)
                .map_err(|e| CodecError::Other(format!("json codec: {e}")))?;
            flat.push(value);
        }

        let shape_u64: Vec<u64> = shape.iter().map(|d| d.get()).collect();
        let text = canonical_to_string(&nest(&flat, &shape_u64));
        Ok(Cow::Owned(text.into_bytes()))
    }

    fn decode<'a>(
        &self,
        bytes: ArrayBytesRaw<'a>,
        shape: &[NonZeroU64],
        data_type: &DataType,
        _fill_value: &FillValue,
        _options: &CodecOptions,
    ) -> Result<ArrayBytes<'a>, CodecError> {
        let element_size = fixed_size(data_type)?;
        let nested: Value = crate::document::strict_from_slice(&bytes)
            .map_err(|e| CodecError::Other(format!("json codec: chunk is not JSON: {e}")))?;

        let shape_u64: Vec<u64> = shape.iter().map(|d| d.get()).collect();
        let mut flat = Vec::new();
        flatten(&nested, &shape_u64, &mut flat)?;

        let type_name = data_type
            .name_v3()
            .map(|n| n.as_ref().to_string())
            .unwrap_or_default();
        let mut out = Vec::with_capacity(flat.len() * element_size);
        for value in flat {
            // SPEC 9.2: a finite JSON number that overflows the target float
            // type's finite range is out of range (a codec error); non-finite
            // values are representable only through the explicit strings.
            // A JSON number, or (for complex) a [re, im] pair of JSON numbers:
            // inputs whose conversion must stay finite (the string forms are
            // the only way to write non-finite values).
            let numeric_input = value.is_number()
                || value
                    .as_array()
                    .is_some_and(|parts| parts.iter().all(Value::is_number));
            let value_text = value.to_string();
            let metadata: FillValueMetadata = serde_json::from_value(value)
                .map_err(|e| CodecError::Other(format!("json codec: {e}")))?;
            let fill_value = data_type
                .fill_value_v3(&metadata)
                .map_err(|e| CodecError::Other(format!("json codec: {e}")))?;
            let element = fill_value.as_ne_bytes();
            if element.len() != element_size {
                return Err(CodecError::Other(format!(
                    "json codec: scalar decoded to {} bytes, expected {element_size}",
                    element.len()
                )));
            }
            if numeric_input && !float_element_is_finite(&type_name, element) {
                return Err(CodecError::Other(format!(
                    "json codec: {type_name} value {value_text} is out of range \
                     (not finite after conversion); use \"NaN\"/\"Infinity\"/\"-Infinity\" \
                     for non-finite values"
                )));
            }
            out.extend_from_slice(element);
        }
        Ok(ArrayBytes::new_flen(out))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use zarrs::array::data_type;

    fn shape(dims: &[u64]) -> Vec<NonZeroU64> {
        dims.iter().map(|d| NonZeroU64::new(*d).unwrap()).collect()
    }

    fn encode_to_string(
        codec: &JsonCodec,
        elements: Vec<u8>,
        chunk_shape: &[NonZeroU64],
        dt: &DataType,
    ) -> String {
        let encoded = codec
            .encode(
                ArrayBytes::new_flen(elements),
                chunk_shape,
                dt,
                &FillValue::new(vec![0; dt.fixed_size().unwrap()]),
                &CodecOptions::default(),
            )
            .unwrap();
        String::from_utf8(encoded.into_owned()).unwrap()
    }

    #[test]
    fn encode_decode_round_trip_uint8_2d() {
        let codec = JsonCodec;
        let dt = data_type::uint8();
        let chunk_shape = shape(&[2, 4]);
        let elements: Vec<u8> = (0..8).collect();
        let text = encode_to_string(&codec, elements.clone(), &chunk_shape, &dt);
        assert_eq!(text, "[[0,1,2,3],[4,5,6,7]]");

        let decoded = codec
            .decode(
                Cow::Owned(text.into_bytes()),
                &chunk_shape,
                &dt,
                &FillValue::new(vec![0]),
                &CodecOptions::default(),
            )
            .unwrap();
        assert_eq!(decoded.into_fixed().unwrap().into_owned(), elements);
    }

    #[test]
    fn decode_rejects_numbers_that_overflow_the_target_float_type() {
        let codec = JsonCodec;
        let big_int = format!("[1{}]", "0".repeat(400)); // integer token, 401 digits
        for (dt, text) in [
            (data_type::float32(), "[1e39]".to_string()),
            (data_type::float64(), big_int),
            (data_type::float16(), "[70000]".to_string()),
            (data_type::complex64(), "[[1e39,0]]".to_string()),
            (data_type::complex128(), format!("[[0,1{}]]", "0".repeat(400))),
        ] {
            let err = codec
                .decode(
                    Cow::Owned(text.clone().into_bytes()),
                    &shape(&[1]),
                    &dt,
                    &FillValue::new(vec![0; dt.fixed_size().unwrap()]),
                    &CodecOptions::default(),
                )
                .err()
                .unwrap_or_else(|| panic!("{text} should be out of range for {dt:?}"));
            // Either our finiteness check or zarrs's own fill-value parsing may
            // reject first; what matters is that it is an error, never Infinity.
            assert!(
                err.to_string().contains("out of range")
                    || err.to_string().contains("incompatible fill value"),
                "{err}"
            );
        }
        // The explicit string form is still how non-finite values are written.
        let ok = codec
            .decode(
                Cow::Owned(b"[\"Infinity\"]".to_vec()),
                &shape(&[1]),
                &data_type::float32(),
                &FillValue::new(vec![0; 4]),
                &CodecOptions::default(),
            )
            .unwrap();
        let v = f32::from_ne_bytes(ok.into_fixed().unwrap().as_ref().try_into().unwrap());
        assert!(v.is_infinite());
    }

    #[test]
    fn encode_float64_nonfinite_uses_fill_value_strings() {
        let codec = JsonCodec;
        let dt = data_type::float64();
        let values = [1.5f64, f64::NAN, f64::INFINITY, f64::NEG_INFINITY, -0.0];
        let elements: Vec<u8> = values.iter().flat_map(|v| v.to_ne_bytes()).collect();
        let chunk_shape = shape(&[5]);
        let text = encode_to_string(&codec, elements.clone(), &chunk_shape, &dt);
        // ES number formatting (RFC 8785): -0.0 prints "0", losing its sign.
        assert_eq!(text, r#"[1.5,"NaN","Infinity","-Infinity",0]"#);

        // Decode reproduces the bit patterns, except -0.0 which comes back
        // as +0.0 (canonical numbers do not preserve the sign of zero).
        let expected: Vec<u8> = [1.5f64, f64::NAN, f64::INFINITY, f64::NEG_INFINITY, 0.0]
            .iter()
            .flat_map(|v| v.to_ne_bytes())
            .collect();
        let decoded = codec
            .decode(
                Cow::Owned(text.into_bytes()),
                &chunk_shape,
                &dt,
                &FillValue::new(vec![0; 8]),
                &CodecOptions::default(),
            )
            .unwrap();
        assert_eq!(decoded.into_fixed().unwrap().into_owned(), expected);
    }

    #[test]
    fn complex_scalars_stop_recursion_at_last_dimension() {
        let codec = JsonCodec;
        let dt = data_type::complex128();
        // One complex element: JSON form is itself a 2-element array.
        let elements: Vec<u8> = [1.5f64, -2.5f64]
            .iter()
            .flat_map(|v| v.to_ne_bytes())
            .collect();
        let chunk_shape = shape(&[1]);
        let text = encode_to_string(&codec, elements.clone(), &chunk_shape, &dt);
        assert_eq!(text, "[[1.5,-2.5]]");

        let decoded = codec
            .decode(
                Cow::Owned(text.into_bytes()),
                &chunk_shape,
                &dt,
                &FillValue::new(vec![0; 16]),
                &CodecOptions::default(),
            )
            .unwrap();
        assert_eq!(decoded.into_fixed().unwrap().into_owned(), elements);
    }

    #[test]
    fn bool_and_int_round_trip() {
        let codec = JsonCodec;
        let dt = data_type::bool();
        let chunk_shape = shape(&[2]);
        let text = encode_to_string(&codec, vec![1, 0], &chunk_shape, &dt);
        assert_eq!(text, "[true,false]");

        let dt = data_type::int32();
        let elements: Vec<u8> = [-5i32, 7].iter().flat_map(|v| v.to_ne_bytes()).collect();
        let text = encode_to_string(&codec, elements, &chunk_shape, &dt);
        assert_eq!(text, "[-5,7]");
    }

    #[test]
    fn decode_shape_mismatch_errors() {
        let codec = JsonCodec;
        let dt = data_type::uint8();
        let result = codec.decode(
            Cow::Borrowed(b"[1,2,3]".as_slice()),
            &shape(&[2]),
            &dt,
            &FillValue::new(vec![0]),
            &CodecOptions::default(),
        );
        assert!(result.is_err());
    }

    #[test]
    fn decode_non_json_errors() {
        let codec = JsonCodec;
        let dt = data_type::uint8();
        let result = codec.decode(
            Cow::Borrowed(b"[NaN]".as_slice()),
            &shape(&[1]),
            &dt,
            &FillValue::new(vec![0]),
            &CodecOptions::default(),
        );
        assert!(result.is_err());
    }

    #[test]
    fn registered_with_zarrs_plugin_registry() {
        let metadata = MetadataV3::new("json");
        let codec = Codec::from_metadata(&metadata).unwrap();
        assert!(matches!(codec, Codec::ArrayToBytes(_)));
    }

    #[test]
    fn nest_and_flatten_are_inverses() {
        let flat: Vec<Value> = (0..24).map(|i| json!(i)).collect();
        for dims in [vec![24], vec![4, 6], vec![2, 3, 4], vec![24, 1]] {
            let nested = nest(&flat, &dims);
            let mut back = Vec::new();
            flatten(&nested, &dims, &mut back).unwrap();
            assert_eq!(back, flat, "shape {dims:?}");
        }
    }
}
